forgevtt-module.js

```diff
@@ -1,2530 +1,2556 @@
+            "THEFORGE.APIRateMonitorSustainedUsageWarning":
+              "Forge API rate monitor warning: {endpoint} on the Forge API has been called continuously for {count} consecutive minutes. Excessive calls may affect performance.",
+            "THEFORGE.APIRateMonitorTroubleshooting": `If you are experiencing poor performance, please check the browser dev tools (F12 or Cmd+Opt+I on Mac). For more information, please see the <a href="https://forums.forge-vtt.com/t/forge-api-rate-monitor/97810#troubleshooting-3" target="_blank">troubleshooting guide</a> or <a href="${ForgeVTT.FORGE_URL}/contact" target="_blank">contact Forge support</a>.`,
+            "THEFORGE.APIRateMonitorLogTrace": `Forge rate monitor: {endpoint} called {calls} times per minute for {consecutive} consecutive minutes. Excessive calls may affect performance.`,
+        };
+    }
+}
+
+ForgeVTT.HEARTBEAT_TIMER = 10 * 60 * 1000; // Send a heartbeat every 10 minutes to update player activity usage and get server updates
+ForgeVTT.ACTIVITY_CHECK_INTERVAL = 15 * 1000; // Check for activity 15 seconds
+ForgeVTT.ACTIVITY_UPDATE_INTERVAL = 60 * 1000; // Update active status every minute
+ForgeVTT.GAME_INACTIVE_THRESHOLD = 2 * 60 * 60 * 1000; // A game inactive for 2 hours should be booted
+ForgeVTT.GAME_SOLO_INACTIVE_THRESHOLD = 1 * 60 * 60 * 1000; // A game inactive for 1 hour with no other players should be booted
+ForgeVTT.OTHER_INACTIVE_THRESHOLD = 50 * 60 * 1000; // A setup/join page inactive for 50 minutes should be booted
+ForgeVTT.HEARTBEAT_ACTIVE_IN_LAST_EVENTS = 10; // Send an active heartbeat if activity detected in the last ACTIVITY_UPDATE_INTERVAL events
+ForgeVTT.IDLE_WARN_ADVANCE = 20 * 60 * 1000;  // Warn the user about being inactive 20 minutes before idling the game
+
+class ForgeVTTPWA extends FormApplication {
+    async render() {
+        const event = this.constructor.installEvent;
+        if (!event) return;
+        event.prompt();
+        const { outcome } = await event.userChoice;
+        if (outcome === "accepted") {
+            ui.notifications.info(`Your Forge game has been installed!`);
+        }
+    }
+}
+/*
+// For testing
+ForgeVTT.HEARTBEAT_TIMER = 1 * 60 * 1000;
+ForgeVTT.ACTIVITY_CHECK_INTERVAL = 10 * 1000;
+ForgeVTT.ACTIVITY_UPDATE_INTERVAL = 60 * 1000;
+ForgeVTT.GAME_INACTIVE_THRESHOLD = 3 * 60 * 1000;
+ForgeVTT.OTHER_INACTIVE_THRESHOLD = 1 * 60 * 1000;
+ForgeVTT.HEARTBEAT_ACTIVE_IN_LAST_EVENTS = 10;
+ForgeVTT.IDLE_WARN_ADVANCE = 60 * 1000;
+*/
+
+class ForgeAPI_RateMonitor {
+    static reset(softReset = false) {
+        this.debug = false; // ForgeAPI_RateMonitor.debug = true to log trace on every call
+        this.timePeriod = 60 * 1000; // 1 minute; If the period changes, please review thresholds and APIRateMonitor messages
+        this.warningFrequency = 10; // Every 10 calls to the same endpoint after a threshold logs another warning
+        this.spikeWarningThreshold = 1000; // >= 1000 calls per minute indicates a spike
```

Comment:
"Spike" wouldn't be the right terminology if it's consistent, but that's just semantics. I really hope no one is hitting the API that hard over a sustained period, haha.

Reply:
That's true, if it's sustained at that level it's not really a spike haha
The var name was based on the patterns observed in testing. Usually, when the rate goes excessively high like that it will eventually be brought back into line by server side rate limiting, at least that is my understanding. I've not seen a sustained spike, but it is technically possible for both the sustained and spike warnings to get thrown side by side, which does cover  that case to an extent. They're not exclusive